blog/en/blog/2023/07/07/tiny-apisix-plugin.md

```diff
@@ -0,0 +1,203 @@
+
+```go title="main.go"
+// references:
+// https://github.com/tetratelabs/proxy-wasm-go-sdk/tree/main/examples
+// https://github.com/apache/apisix/blob/master/t/wasm/
+package main
+
+import (
+    "github.com/tetratelabs/proxy-wasm-go-sdk/proxywasm"
+    "github.com/tetratelabs/proxy-wasm-go-sdk/proxywasm/types"
+
+    "github.com/valyala/fastjson"
+)
+
+func main() {
+    proxywasm.SetVMContext(&vmContext{})
+}
+
+// each plugin has its own VMContext.
+// it is responsible for creating multiple PluginContexts for each route.
+type vmContext struct {
+    types.DefaultVMContext
+}
+
+// each route has its own PluginContext.
+// it corresponds to one instance of the plugin.
+func (*vmContext) NewPluginContext(contextID uint32) types.PluginContext {
+    return &pluginContext{}
+}
+
+type header struct {
+    Name  string
+    Value string
+}
+
+type pluginContext struct {
+    types.DefaultPluginContext
+    Headers []header
+}
+
+func (ctx *pluginContext) OnPluginStart(pluginConfigurationSize int) types.OnPluginStartStatus {
+    data, err := proxywasm.GetPluginConfiguration()
+    if err != nil {
+        proxywasm.LogErrorf("error reading plugin configuration: %v", err)
+        return types.OnPluginStartStatusFailed
+    }
+
+    var p fastjson.Parser
+    v, err := p.ParseBytes(data)
+    if err != nil {
+        proxywasm.LogErrorf("error decoding plugin configuration: %v", err)
+        return types.OnPluginStartStatusFailed
+    }
+    headers := v.GetArray("headers")
+    ctx.Headers = make([]header, len(headers))
+    for i, hdr := range headers {
+        ctx.Headers[i] = header{
+            Name:  string(hdr.GetStringBytes("name")),
+            Value: string(hdr.GetStringBytes("value")),
+        }
+    }
+    return types.OnPluginStartStatusOK
+}
+
+// each HTTP request to a route has its own HTTPContext
+func (ctx *pluginContext) NewHttpContext(contextID uint32) types.HttpContext {
+    return &httpContext{parent: ctx}
+}
+
+type httpContext struct {
+    types.DefaultHttpContext
+    parent *pluginContext
+}
+
+func (ctx *httpContext) OnHttpResponseHeaders(numHeaders int, endOfStream bool) types.Action {
+    plugin := ctx.parent
+    for _, hdr := range plugin.Headers {
+        proxywasm.ReplaceHttpResponseHeader(hdr.Name, hdr.Value)
+    }
+
+    return types.ActionContinue
+}
+```
+
+To compile our plugin to a Wasm binary, we can run:
+
+```shell
+tinygo build -o custom_response_header.go.wasm -scheduler=none -target=wasi ./main.go
+```
+
+## Configuring APISIX to Run the Plugin
+
+To use the Wasm plugin, we first have to update our APISIX configuration file to add this:
+
+```yaml title="config.yaml"
+wasm:
+  plugins:
+    - name: custom-response-header
+      priority: 7000
+      file: /opt/apisix/wasm/custom_response_header.go.wasm
+```
+
+Now we can create a route and enable this plugin:
```

Comment:
If a new user comes to this blog, they won't know that the following yaml configuration requires APISIX to be running on standalone mode. A little description about getting started with APISIX or a prerequisite list or a hyperlink to all necessary resources would be very neat.